otaclient/app/common.py

```diff
@@ -336,264 +337,190 @@ def urljoin_ensure_base(base: str, url: str):
+        max_concurrent: int,
+        backoff_func: Callable[[int], float],
+    ) -> None:
+        # task dispatch interval for continues failling
+        self.started = False  # can only be started once
+        self._backoff_func = backoff_func
+        self._executor = executor
+        self._shutdown_event = threading.Event()
+        self._se = threading.Semaphore(max_concurrent)
 
-class _ThreadSafeGenWrapper:
-    def __init__(self, _gen: Generator, *, kick_start: bool) -> None:
-        self._gen = _gen
-        self._lock = Lock()
-        if kick_start:
-            self.send(None)
+        self._total_tasks_count = 0
+        self._dispatched_tasks: Set[Future] = set()
+        self._failed_tasks: Set[T] = set()
+        self._last_failed_fut: Optional[Future] = None
 
-    def close(self):
-        self._gen.close()
+        self._done_task_counter = itertools.count(start=1)
+        self._all_done = threading.Event()
+        self._dispatch_done = False
 
-    def send(self, __value: Any) -> Any:
-        try:
-            with self._lock:
-                return self._gen.send(__value)
-        except StopIteration:
-            pass  # ignore on closed gen
+        self._done_que: Queue[DoneTask] = Queue()
 
-    def throw(self, __exp: Exception):
-        try:
-            with self._lock:
-                self._gen.throw(__exp)
-        except Exception as __exp:  # break ref cycle
-            raise __exp
+    def _done_task_cb(self, item: T, fut: Future):
+        """
+        Tracking done counting, set all_done event.
+        add failed to failed list.
+        """
+        self._se.release()  # always release se first
+        # NOTE: don't change dispatched_tasks if shutdown_event is set
+        if self._shutdown_event.is_set():
+            return
+
+        self._dispatched_tasks.discard(fut)
+        # check if we finish all tasks
+        _done_task_num = next(self._done_task_counter)
+        if self._dispatch_done and _done_task_num == self._total_tasks_count:
+            logger.debug("all done!")
+            self._all_done.set()
+
+        if fut.exception():
+            self._failed_tasks.add(item)
+            self._last_failed_fut = fut
+        self._done_que.put_nowait(DoneTask(fut, item))
+
+    def _task_dispatcher(self, func: Callable[[T], Any], _iter: Iterable[T]):
+        """A dispatcher in a dedicated thread that dispatches
+        tasks to threadpool."""
+        for item in _iter:
+            if self._shutdown_event.is_set():
+                return
+            self._se.acquire()
+            self._total_tasks_count += 1
+
+            fut = self._executor.submit(func, item)
+            fut.add_done_callback(partial(self._done_task_cb, item))
+            self._dispatched_tasks.add(fut)
+        logger.debug(f"dispatcher done: {self._total_tasks_count=}")
+        self._dispatch_done = True
+
+    def _done_task_collecter(self) -> Generator[DoneTask, None, None]:
```

Comment:
collecter -> collector ?

Reply:
Fixed in [7dd1586](https://github.com/tier4/ota-client/pull/232/commits/7dd1586a6a38c54facb6d8e77e861559bb6cd4b1).